[TECH] Utiliser le système de module ESM (PIX-7202)

octo-topi opened the pull request:
## :unicorn: Problème
Actuellement, l'api de pix utilise les modules CommonJs. 
Ce n'est pas une solution compatible pour la migration vers typescript.

## :robot: Proposition
Utiliser le système de module ESM.

## :rainbow: Remarques
Utilisation de codemod et modifications manuelles.

## :100: Pour tester
Relire l'ADR.
Vérifier la CI.
Faire des tests manuels en RA.

matthiasferraina committed:
feat(api): check all modules are esm

Co-authored-by: Pierre TOP (TOPI) <ptop@octo.com>

octo-topi committed:
Add named export codemod

matthiasferraina committed:
feat(api): create codemod for functions inside named export

matthiasferraina committed:
feat(api): create codemod for anonymous export from identifier to named export

matthiasferraina committed:
feat(api): add codemod to migrate bookshelf cjs anonymous call expression exports to cjs named exports

matthiasferraina committed:
feat(api): add codemod to migrate joi cjs anonymous call expression exports to cjs named exports

matthiasferraina committed:
feat(api): add codemod to migrate cjs class anonymous exports to cjs named exports

octo-topi committed:
Create CJS export to ESM export

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr5787.review.pix.fr)
  * [App (.org)](https://app-pr5787.review.pix.org)
  * [Orga](https://orga-pr5787.review.pix.fr)
  * [Certif](https://certif-pr5787.review.pix.fr)
  * [Admin](https://admin-pr5787.review.pix.fr)
  * [API](https://api-pr5787.review.pix.fr/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr5787/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr5787/environment)

matthiasferraina committed:
Create CJS to ESM named import codemod

matthiasferraina committed:
feat(api): add eslint-nibble to monitor migration

matthiasferraina committed:
feat(api): create codemod for cjs named import using require with a key

matthiasferraina committed:
faet(api): create codemod for cjs export directly specified on exports expression

matthiasferraina committed:
feat(api): create codemod to use namespace import for modules with callExpression related to import

matthiasferraina committed:
feat(api): create codemod for package import

matthiasferraina committed:
feat(api): create codemod for import invocation only

matthiasferraina committed:
chore(api): manually fix invalid imports

octo-topi committed:
chore(api): manually fix invalid imports on classes (domain and orm)

matthiasferraina committed:
feat(api): fix config esm import

matthiasferraina committed:
chore(api): manually fix invalid imports part two

matthiasferraina committed:
chore(api): manually fix invalid imports on joi validators

matthiasferraina committed:
chore(api): manually fix invalid imports on builders

octo-topi committed:
Remove unused json bookshelf plugin

octo-topi committed:
Add exception for process.exit in worker

octo-topi committed:
Add index.js

octo-topi committed:
Import JSON files

matthiasferraina committed:
chore(api): refactor to avoid js reserved keyword delete

matthiasferraina committed:
chore(api): fix no-undef lint rule

matthiasferraina committed:
chore(api): fix no-restricted-syntax lint rule

matthiasferraina committed:
chore(api): fix __filename and dirname violations

matthiasferraina committed:
chore(api): fix no-undef rule

matthiasferraina committed:
feat(api): create codemod for  cjs lib import in esm

matthiasferraina committed:
feat(api): add eslint plugin to parse json import

octo-topi committed:
Prepare lodash import for package codemod

octo-topi committed:
Specify import extension on JSON files

octo-topi committed:
Specify filename extension on dayjs plugins

octo-topi committed:
Fix knex invalid configuration when using ESM

octo-topi committed:
Disable lint violation on dynamic import

octo-topi committed:
Fix implicit index in pdf-lib package

octo-topi committed:
Add missing file extension

octo-topi committed:
Run package codemod

octo-topi committed:
Configure eslint-nibble

octo-topi committed:
Exclude nyc config from lint

octo-topi committed:
Fix allowed named import in fs and pdf-lib packages

octo-topi committed:
Fix allowed named import in lodash packages

matthiasferraina committed:
chore(api): fix package imports

matthiasferraina committed:
chore(api): fix worker

matthiasferraina committed:
chore(api): Fix allowed named import in uuid package

matthiasferraina committed:
Disable lint violation on dynamic import

matthiasferraina committed:
chore(api): Fix allowed named import in json2csv package

matthiasferraina committed:
chore(api): Fix allowed named import in jsonapiSerializer package

octo-topi committed:
Fix allowed named import in hapi package

octo-topi committed:
Break dependencies between controllers and scripts

octo-topi committed:
Fix anonymous import of named export

matthiasferraina committed:
chore(api): remove uncompatible job name

matthiasferraina committed:
chore(api): fix eventBusBuilder invocation

matthiasferraina committed:
chore(api): fix sql correlation with http call

matthiasferraina committed:
fix(api): make dotenv import explicit in config.js

matthiasferraina committed:
feat(api): run lint and codemod on all codebase

matthiasferraina committed:
chore(api): fix fatal linting errors

octo-topi committed:
Fix untriggerred existing lint violations

matthiasferraina committed:
chore(api): fix config imports

matthiasferraina committed:
chore(api): fix usecases import

octo-topi committed:
Fix constants using namespaced import

octo-topi committed:
Fix oidc-identity-provider using namespaced import

octo-topi committed:
Fix CertificationIssueReportCategory using namespaced import

matthiasferraina committed:
chore(api): fix no-missing-import violations

octo-topi committed:
Fix untrigerred existing eslint-comment violations

octo-topi committed:
Fix import-default violations

octo-topi committed:
Fix no-unused-var violation

octo-topi committed:
Fix untrigerred existing no-empty-function

octo-topi committed:
Fix no-import-assign violation

octo-topi committed:
Fix named exports in repository

matthiasferraina committed:
chore(api): fix __dirname import

matthiasferraina committed:
chore(api): manually migrate test-helper file

octo-topi committed:
Remove unused strict clause

octo-topi committed:
Manually fix codemod failure on require

octo-topi committed:
Fix dotenv import in scripts

octo-topi committed:
Manually fix codemod failure on module.exports

matthiasferraina committed:
chore(api): fix import-named violation

octo-topi committed:
Fix CLI conditional launch

octo-topi committed:
Substitute __filename pseudo-variable

octo-topi committed:
Fix no-unused-vars, import/default

octo-topi committed:
Remove import mutation in certification statistics

This operation is allowed is CJS but not in ESM
Either use a proxy or implement cache in repository

matthiasferraina committed:
chore(api): fix lodash import

matthiasferraina committed:
chore(api): fix knex

matthiasferraina committed:
chore(api): add js file extension

matthiasferraina committed:
fix(api): fix script invocation in test

matthiasferraina committed:
Fix aliased export on certification seeds

octo-topi committed:
Show how DI would solve this horrible mess

octo-topi committed:
Inject dependencies in unit tests : controller

matthiasferraina committed:
chore(api): fix temporary storage import

matthiasferraina committed:
chore(api): inject dependencies in unit tests : domain/models

VincentHardouin committed:
chore(api): inject dependencies in admin-member-controller

Co-authored-by: Jérémy Pluquet <jeremy.pluquet@pix.fr>

La-toile-cosmique committed:
feat(api): add dependency injection in target profile controller

francois2metz commented on api/index.js:
```diff
@@ -46,9 +46,6 @@ process.on('SIGINT', () => {
 (async () => {
   try {
     await start();
-    if (process.env.START_JOB_IN_WEB_PROCESS) {
```
On veut vraiment faire ça ?

matthiasferraina replied:
On soiuhaiterait plutot s'en passer effectivement . On avait étudié ce genre d'alternative : https://github.com/1024pix/pix/pull/5808/files

octo-topi replied:
C'est corrigé avec un import dynamique :rocket:

matthiasferraina committed:
chore(api): use wrapper for router unit test to fix sinon issue

matthiasferraina committed:
chore(api): fix learningContentBuilder invocation

matthiasferraina committed:
chore(api): export lcms wrapper to fix sinon issue

matthiasferraina committed:
chore(api): fixes

octo-topi committed:
Sadly add exception to dependency check

octo-topi committed:
Remove proxyquire dependency

octo-topi committed:
Switch back database lint to CJS

matthiasferraina committed:
chore(api): inject dependencies in create-user-orga-settings test

octo-topi committed:
Export function in repository rather than object

matthiasferraina committed:
refacto(api): remove unecessary test

matthiasferraina committed:
chore(api): inject dependencies in generate-knowledge-element script

octo-topi committed:
Make test deterministic on repository test

octo-topi committed:
Use named import for ToS use-cases

octo-topi committed:
Use named export for repositories

octo-topi committed:
Fix invalid getter translation to ESM

octo-topi committed:
Fix flaky test due to unsorted array

matthiasferraina committed:
refactor(api): delete tests that are not unit tests

octo-topi committed:
Inject configuration in Scorecard model

octo-topi committed:
Replace no-longer compatible class assertion by value in authentication service registry

octo-topi committed:
Inject dependecy in jury session serializer

octo-topi committed:
Inject dependecy in organization invitation controller

octo-topi committed:
Inject serializer dependecy in organization invitation controller

octo-topi committed:
Inject certification pdf util dependecy in organization controller

octo-topi committed:
Inject mail service dependecy in organization invitation controller

matthiasferraina committed:
refactor(api): restore actual implementation for transitive repository dependency

matthiasferraina committed:
inject dependency in campaign-repository

matthiasferraina committed:
inject dependency in send-shared-participation-results-to-pole-emploi_test

matthiasferraina committed:
refactor(api): final fixes !

matthiasferraina committed:
test to fix e2e job

matthiasferraina committed:
chore(api): fix algo test

matthiasferraina committed:
sr(api): leave no trace behind

octo-topi committed:
Document decision

laura-bergoens commented:
Bon pour moi, j'ai évidemment pas tout regardé mais j'ai lu l'ADR, quelques fichiers pour voir ce que ça donnait et je fais confiance aux tests.
En revanche, j'ai pull la branche chez moi pour regarder un peu et je n'arrive pas à configurer mon IDE correctement pour qu'il arrête de me sortir une erreur, une idée ? (j'ai essayé manual et auto pour la config, les deux me laissent l'erreur)
![error_ide_1](https://user-images.githubusercontent.com/48727874/236441081-f0e5745e-0b5e-41a2-8b4c-20f87d8997c7.png)

octo-topi commented:
> Bon pour moi, j'ai évidemment pas tout regardé mais j'ai lu l'ADR, quelques fichiers pour voir ce que ça donnait et je fais confiance aux tests. En revanche, j'ai pull la branche chez moi pour regarder un peu et je n'arrive pas à configurer mon IDE correctement pour qu'il arrête de me sortir une erreur, une idée ? (j'ai essayé manual et auto pour la config, les deux me laissent l'erreur) ![error_ide_1](https://user-images.githubusercontent.com/48727874/236441081-f0e5745e-0b5e-41a2-8b4c-20f87d8997c7.png)

As-tu installé les librairies `npm install` ? Y'a plein de plugins en plus, c'est pour cela que l'erreur poppe IMHO

octo-topi committed:
Inject dependency in CPF job

VincentHardouin commented:
Afin de faciliter la review est-ce que les premiers commits qui sont des codemods ne peuvent pas être dans une PR spécifique de préparation et garder cette PR pour uniquement la migration ? 

edit: Et si c'est possible pour les injections aussi 🫣

VincentHardouin commented on api/worker.js:
```diff
@@ -34,7 +34,7 @@ async function runJobs() {
   process.on('SIGINT', async () => {
     await monitoredJobQueue.stop();
-    // eslint-disable-next-line node/no-process-exit
+    // eslint-disable-next-line node/no-process-exit,no-process-exit
```
Pourquoi il faut désactiver deux fois la règle ? une première fois avec node et l'autre sans node

VincentHardouin commented on api/translations/index.js:
```diff
@@ -1,4 +1,4 @@
-import { fr } from './fr.js';
-import { en } from './en.js';
+import * as fr from "./fr.json";
```
Les doubles quotes sont ok dans ce fichier ? 🤔

VincentHardouin commented on api/lib/application/sessions/index.js:
```diff
@@ -720,7 +720,7 @@ const register = async function (server) {
           },
         ],
-        handler: sessionController.delete,
+        handler: sessionController.remove,
```
Je vois pas le renommage du controller associé dans [ce commit](https://github.com/1024pix/pix/commit/da8a9814b94c89284b00969c995947884346f077) ? 🤔

matthiasferraina committed:
chore(api): run codemod

matthiasferraina committed:
chore(api): run codemod on others

matthiasferraina committed:
chore(api): fix linting errors for seeds

matthiasferraina committed:
fix(api): fix participantResultsSharedRepository import to repair dependency injection

octo-topi committed:
Remove no longer needed explicit strict mode

francois2metz reviewed (commented):
:rose: :hibiscus: :white_flower: :wilted_flower:  :tulip: :sunflower: :cherry_blossom:  LGTM :cherry_blossom: :sunflower:  :tulip: :wilted_flower: :white_flower: :hibiscus: :rose:

mickaelalibert reviewed (approved):
ça marche en prod 🦄
